File: graphs/399-evaluate-division/EvaluateDivision.py
class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        graph = defaultdict(dict)

        # Build graph
        for (a, b), val in zip(equations, values):
            graph[a][b] = val
            graph[b][a] = 1 / val

        def dfs(curr, target, product, visited):
            if curr == target:
                return product
            visited.add(curr)
            for neighbor, val in graph[curr].items():
                if neighbor not in visited:
                    res = dfs(neighbor, target, product * val, visited)
                    if res != -1:
                        return res
            return -1

        results = []
        for start, end in queries:
            if start not in graph or end not in graph:
                results.append(-1.0)
            else:
                results.append(dfs(start, end, 1, set()))
        return results
        

File: graphs/399-evaluate-division/EvaluateDivision.java
class Solution {
    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
        Map<String, Map<String, Double>> graph = buildGraph(equations, values);
        double[] result = new double[queries.size()];

        for (int i = 0; i < queries.size(); i++) {
            String start = queries.get(i).get(0);
            String end = queries.get(i).get(1);
            if (!graph.containsKey(start) || !graph.containsKey(end)) {
                result[i] = -1.0;
            } else if (start.equals(end)) {
                result[i] = 1.0;
            } else {
                result[i] = dfs(graph, start, end, new HashSet<>(), 1.0);
            }
        }

        return result;
    }

    private double dfs(Map<String, Map<String, Double>> graph, String curr, String target, Set<String> visited, double product) {
        visited.add(curr);
        Map<String, Double> neighbors = graph.get(curr);

        if (neighbors.containsKey(target)) {
            return product * neighbors.get(target);
        }

        for (String next : neighbors.keySet()) {
            if (!visited.contains(next)) {
                double res = dfs(graph, next, target, visited, product * neighbors.get(next));
                if (res != -1.0) {
                    return res;
                }
            }
        }

        return -1.0;
    }

    private Map<String, Map<String, Double>> buildGraph(List<List<String>> equations, double[] values) {
        Map<String, Map<String, Double>> graph = new HashMap<>();

        for (int i = 0; i < equations.size(); i++) {
            String a = equations.get(i).get(0);
            String b = equations.get(i).get(1);
            double val = values[i];

            graph.putIfAbsent(a, new HashMap<>());
            graph.putIfAbsent(b, new HashMap<>());

            graph.get(a).put(b, val);
            graph.get(b).put(a, 1 / val);
        }

        return graph;
    }
}